MoreLinq.Test/TrySingleTest.cs

```diff
@@ -0,0 +1,193 @@
+        [TestCase(SourceKind.BreakingCollection)]
+        [TestCase(SourceKind.BreakingReadOnlyCollection)]
+        public void TrySingleWithMoreThanOne(SourceKind kind)
+        {
+            var arrayWithMultiple = new int?[] { 10, 20 }.ToSourceKind(kind);
+
+            var (cardinality, value) = arrayWithMultiple.TrySingle("zero", "one", "many");
+
+            Assert.AreEqual(cardinality, "many");
+            Assert.IsNull(value);
+        }
+
+        [Test]
+        public void TrySingleDoesNotConsumeMoreThanTwoElementsFromTheSequence()
+        {
+            IEnumerable<int> TestSequence()
+            {
+                yield return 1;
+                yield return 2;
+                throw new Exception(nameof(MoreEnumerable.TrySingle) + " should not have attempted to consume a third element.");
+            }
+
+            var (cardinality, value) = TestSequence().TrySingle("zero", "one", "many");
+            Assert.AreEqual("many", cardinality);
+            Assert.AreEqual(default(int), value);
+        }
+
+        [Test]
+        public void TrySingleOptimizesForList()
+        {
+            var list = new BreakingList<int>();
+
+            var (cardinality, value) = list.TrySingle("zero", "one", "many");
+            Assert.AreEqual("zero", cardinality);
+            Assert.AreEqual(default(int), value);
+
+            list = new BreakingList<int>(new List<int> { 1 });
+            (cardinality, value) = list.TrySingle("zero", "one", "many");
+            Assert.AreEqual("one", cardinality);
+            Assert.AreEqual(1, value);
+
+            list = new BreakingList<int>(new List<int> { 1, 2 });
+            (cardinality, value) = list.TrySingle("zero", "one", "many");
+            Assert.AreEqual("many", cardinality);
+            Assert.AreEqual(default(int), value);
+        }
+
+        [Test]
+        public void TrySingleOptimizesForCollection()
+        {
+            var coll = new BreakingCollection<int>();
+
+            var (cardinality, value) = coll.TrySingle("zero", "one", "many");
+            Assert.AreEqual("zero", cardinality);
+            Assert.AreEqual(default(int), value);
+
+            coll = new BreakingCollection<int>(new List<int> {1, 2});
+
+            (cardinality, value) = coll.TrySingle("zero", "one", "many");
+            Assert.AreEqual("many", cardinality);
+            Assert.AreEqual(default(int), value);
+        }
+
+        [TestCase(0, "zero")]
+        [TestCase(1, "one")]
+        [TestCase(2, "many")]
+        public void TrySingleEnumeratesOnceOnlyAndDisposes(int numberOfElements, string expectedCardinality)
+        {
+            using (var seq = Enumerable.Range(1, numberOfElements).AsTestingSequence())
+            {
+                var (cardinality, _) = seq.TrySingle("zero", "one", "many");
+                Assert.AreEqual(expectedCardinality, cardinality);
+            }
+        }
+
+        private class BreakingSingleElementCollection<T> : ICollection<T>
```

Comment:
```suggestion
        sealed class BreakingSingleElementCollection<T> : ICollection<T>
```

Reply:
Also, this can share its implementation with `BreakingSingleElementReadOnlyCollection` via a base `BreakingSingleElementSequence` to avoid redundancies. Here's what I am thinking:

```c#
class BreakingSingleElementCollectionBase<T> : IEnumerable<T>
{
    readonly T _element;

    protected BreakingSingleElementCollectionBase(T element) => _element = element;

    public int Count => 1;

    public IEnumerator<T> GetEnumerator()
    {
        yield return _element;
        throw new Exception($"{nameof(MoreEnumerable.TrySingle)} should not have attempted to consume a second element.");
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}

sealed class BreakingSingleElementCollection<T> :
    BreakingSingleElementCollectionBase<T>, ICollection<T>
{
    public BreakingSingleElementCollection(T element) : base(element) {}

    public void Add(T item) => throw new NotImplementedException();
    public void Clear() => throw new NotImplementedException();
    public bool Contains(T item) => throw new NotImplementedException();
    public void CopyTo(T[] array, int arrayIndex) => throw new NotImplementedException();
    public bool Remove(T item) => throw new NotImplementedException();
    public bool IsReadOnly => true;
}

sealed class BreakingSingleElementReadOnlyCollection<T> :
    BreakingSingleElementCollectionBase<T>, IReadOnlyCollection<T>
{
    public BreakingSingleElementReadOnlyCollection(T element) : base(element) {}
}
```